man/man1/gfrepcheck.1
'\" t
.\"     Title: gfrepcheck
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 26 Oct 2017
.\"    Manual: Gfarm
.\"    Source: Gfarm
.\"  Language: English
.\"
.TH "GFREPCHECK" "1" "26 Oct 2017" "Gfarm" "Gfarm"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
gfrepcheck \- control replica_check
.SH "SYNOPSIS"
.HP \w'\fBgfrepcheck\fR\ 'u
\fBgfrepcheck\fR [\-P\ \fIpath\fR] {start(or\ enable) | stop(or\ disable) | status}
.HP \w'\fBgfrepcheck\fR\ 'u
\fBgfrepcheck\fR [\-P\ \fIpath\fR] {remove | reduced_log | remove_grace_used_space_ratio | remove_grace_time | host_down_thresh | sleep_time | minimum_interval} {\fIVALUE\fR | status}
.SH "DESCRIPTION"
.PP
Displays the current settings and modifies the settings for replica_check\&.
.PP
When "status" is specified as an argument, the current settings are displayed\&. "gfrepcheck status" displays not only the current setting but also the current status of processing\&.
.PP
When a value for the configuration directive is specified as an argument, the setting is dynamically changed\&. gfarm2\&.conf should be edited to persist the settings\&.
.PP
Please see the manual page of gfarm2\&.conf(5) for details of the following configuration directives\&.
.PP
gfrepcheck * : "replica_check" directive
.PP
gfrepcheck remove * : "replica_check_remove" directive
.PP
gfrepcheck reduced_log * : "replica_check_reduced_log" directive
.PP
gfrepcheck remove_grace_used_space_ratio * : "replica_check_remove_grace_used_space_ratio" directive
.PP
gfrepcheck remove_grace_time * : "replica_check_remove_grace_time" directive
.PP
gfrepcheck host_down_thresh * : "replica_check_host_down_thresh" directive
.PP
gfrepcheck sleep_time * : "replica_check_sleep_time" directive
.PP
gfrepcheck minimum_interval * : "replica_check_minimum_interval" directive
.SH "OPTIONS"
.PP
\fB\-P\fR \fIpath\fR
.RS 4
This option specifies a pathname or a Gfarm URL to identify a metadata server which will be accessed by this command\&. If not specified, "/" is used as the default value\&.
.RE
.PP
\fB\-?\fR
.RS 4
Prints usage\&.
.RE


ChangeLog
2007-03-31  Noriyuki Soda  <soda@sra.co.jp>

	* RELNOTES
	avoid an expression which may be misinterpreted as a security problem.

2007-03-31  Osamu Tatebe <tatebe@cs.tsukuba.ac.jp>

	* RELNOTES
	correct description

2007-03-30  Tohru Sotoyama <sotoyama@sra.co.jp>

	* package/solaris/build
	change globus root directory

2007-03-30  Noriyuki Soda  <soda@sra.co.jp>

	* ChangeLog LICENSE RELNOTES
	merge changes in the CVS main trunk until gfarm-1.4.1 to the GNEO
	branch
	
	* ChangeLog
	Catched up with the current ChangeLog.

	* ChangeLog
	recover missed log messages in previous commit
	
	* doc/html/en/ref/man5/gfarm.conf.5.html, doc/html/ja/ref/man5/gfarm.conf.5.html man/ja/man5/gfarm.conf.5 man/man5/gfarm.conf.5:
	regen

	* doc/Gfarm-FAQ.en, doc/Gfarm-FAQ.ja, doc/docbook/en/ref/man5/gfarm.conf.5.docbook, doc/docbook/ja/ref/man5/gfarm.conf.5.docbook:
	merge the following changes in the SourceForge Subversion repository:

	Revision: 3674
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3674&view=rev
	Author:   sotoyama
	Date:     2007-03-29 02:34:39 -0700 (Thu, 29 Mar 2007)
	Log Message:
	add limitation about mixing gsi authentication with sharedsecret
	found in [gfarm-developers:02814]

	Revision: 3676
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3676&view=rev
	Author:   sotoyama
	Date:     2007-03-29 04:46:11 -0700 (Thu, 29 Mar 2007)
	Log Message:
	add limitation about mixing gsi authentication with sharedsecret
	found in [gfarm-developers:02814] to `auth' statement

	Revision: 3682
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3682&view=rev
	Author:   n-soda
	Date:     2007-03-29 11:03:31 -0700 (Thu, 29 Mar 2007)
	Log Message:
	swap section order of limitation and example of the "auth" directive,
	and add a paragraph to describe the example.

2007-03-30  yusuke Tanimura <yusuke.tanimura@aist.go.jp>

	* RELNOTES:
	proofreading

2007-03-29  Noriyuki Soda  <soda@sra.co.jp>

	* server/gfsd/gfsd.c:
	merge the follownig change from the SourceForge Subversion repository:

	Revision: 3681
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3681&view=rev
	Author:   n-soda
	Date:     2007-03-29 07:37:36 -0700 (Thu, 29 Mar 2007)
	Log Message:
	report error status of close(2) in gfs_server_open_local(),
	problem reported by Markus Elfring as:
	http://sourceforge.net/tracker/index.php?func=detail&aid=1687850&group_id=172718&atid=862742

	Also, treat an error at sendmsg(2) as just an error instead of
	fatal one.

	* bench/fsysbench/fsysbench.c:
	merge the follownig change from the SourceForge Subversion repository:

	Revision: 3679
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3679&view=rev
	Author:   n-soda
	Date:     2007-03-29 05:13:05 -0700 (Thu, 29 Mar 2007)
	Log Message:
	timer_interrupt_handler() shouldn't call printf(3), but write(2),
	because the former is not async-signal-safe.
	c.f.
	https://sourceforge.net/tracker/?func=detail&atid=862742&aid=1687842&group_id=172718
	
	* ftool/gfarm_agent/gfarm_agent.c:
	merge the follownig change from the SourceForge Subversion repository:

	Revision: 3678
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3678&view=rev
	Author:   n-soda
	Date:     2007-03-29 04:59:54 -0700 (Thu, 29 Mar 2007)
	Log Message:
	sigterm_handler() shouldn't call exit(3), but _exit(2), because
	the former is not async-signal-safe.
	c.f.
	https://sourceforge.net/tracker/?func=detail&atid=862742&aid=1687842&group_id=172718
	
	* server/gfsd/gfsd.c:
	merge the follownig change from the SourceForge Subversion repository:

	Revision: 3675
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3675&view=rev
	Author:   n-soda
	Date:     2007-03-29 03:30:02 -0700 (Thu, 29 Mar 2007)
	Log Message:
	do not call exit(3) but _exit(2) in accepting_sigterm_handler(),
	because the former is not async-signal-safe.
	problem reported by Markus Elfring in
	http://sourceforge.net/tracker/index.php?func=detail&aid=1687842&group_id=172718&atid=862742

2007-03-29  yusuke Tanimura <yusuke.tanimura@aist.go.jp>

	* RELNOTES, LICENSE:
	update release notes and the copyright display for the version 1.4.1

2007-03-28  Osamu Tatebe <tatebe@cs.tsukuba.ac.jp>

	* ChangeLog:
	Catched up with the current ChangeLog.

2007-03-28  Noriyuki Soda  <soda@sra.co.jp>

	* include/gfarm/gfarm_misc.h, package/freebsd/Makefile,	package/pkgsrc/Makefile, package/redhat/gfarm.spec, package/solaris/build:
	increase version number to 1.4.1

	* include/gfarm/gfarm_error.h, lib/libgfarm/gfarm/error.c, lib/libgfarm/gfarm/gfs_client.c:
	merge the following change from the SourceForge Subversion repository:

	Revision: 3673
		  http://gfarm.svn.sourceforge.net/gfarm/?rev=3673&view=rev
	Author:   n-soda
	Date:     2007-03-28 07:11:37 -0700 (Wed, 28 Mar 2007)
	Log Message:
	make gfsd reconnection feature work against more network errors.
	this fixes a problem reported by Yamamoto Naotaka-san
	in [gfarm-developers:03456] which was caused by ECONNRESET.

	introduce ENETDOWN, EHOSTDOWN and ENETRESET as gfarm native errors
	for this purpose. (although ENETRESET will be nearly never used)

	* gftool/gneo/gneo.c:
	just a portability fix

	* regress/gfs_hook/lib/fsystest/index.FreeBSD.expected, regress/gfs_hook/lib/fsystest/index.Linux.expected, regress/gfs_hook/lib/fsystest/index.NetBSD.expected, regress/gfs_hook/lib/fsystest/index.SunOS.expected:
	results are different on the GNEO branch with gfs_hook


gftool/gfrepcheck/gfrepcheck.c
#include <stdlib.h>
#include <unistd.h>
#include <libgen.h>
#include <stdio.h>
#include <string.h>
#include <gfarm/gfarm.h>

#include "config.h"
#include "gfm_client.h"
#include "lookup.h"
#include "gfarm_path.h"

char *program_name = "gfrepcheck";

static void
error_check(const char *msg, gfarm_error_t e)
{
	if (e == GFARM_ERR_NO_ERROR)
		return;

	fprintf(stderr, "%s: %s\n", msg, gfarm_error_string(e));
	exit(EXIT_FAILURE);
}

static void
usage(void)
{
	fprintf(stderr, "Usage: \t%s %s\n" "\t%s %s\n" "\t%s %s\n"
	    "\t%s %s\n" "\t%s %s\n" "\t%s %s\n"
	    "\t%s %s\n" "\t%s %s\n" "\t%s %s\n"
	    "\t%s %s\n" "\t%s %s\n" "\t%s %s\n" "\t%s %s\n"
	    "\t%s %s\n" "\t%s %s\n" "\t%s %s\n" "\t%s %s\n"
	    "\t%s %s\n" "\t%s %s\n",
	    program_name, "[-P <path>] status",
	    program_name, "[-P <path>] start(or enable)",
	    program_name, "[-P <path>] stop(or disable)",
	    program_name, "[-P <path>] remove status",
	    program_name, "[-P <path>] remove enable",
	    program_name, "[-P <path>] remove disable",
	    program_name, "[-P <path>] reduced_log status",
	    program_name, "[-P <path>] reduced_log enable",
	    program_name, "[-P <path>] reduced_log disable",
	    program_name, "[-P <path>] remove_grace_used_space_ratio status",
	    program_name, "[-P <path>] remove_grace_used_space_ratio PERCENT",
	    program_name, "[-P <path>] remove_grace_time status",
	    program_name, "[-P <path>] remove_grace_time SECOND",
	    program_name, "[-P <path>] host_down_thresh status",
	    program_name, "[-P <path>] host_down_thresh SECOND",
	    program_name, "[-P <path>] sleep_time status",
	    program_name, "[-P <path>] sleep_time NANOSECOND",
	    program_name, "[-P <path>] minimum_interval status",
	    program_name, "[-P <path>] minimum_interval SECOND");
	exit(EXIT_FAILURE);
}

enum target { START, STOP, STATUS1,
	      REMOVE, REDUCED_LOG,
	      REMOVE_GRACE_USED_SPACE_RATIO, REMOVE_GRACE_TIME,
	      HOST_DOWN_THRESH, SLEEP_TIME, MINIMUM_INTERVAL };
enum value { ENABLE, DISABLE, STATUS2, SET_VALUE };

#define BUFSIZE 2048

static gfarm_error_t
config_get(struct gfm_connection *gfm_server, enum target target)
{
	gfarm_error_t e;
	char buffer[BUFSIZE];
	char config[BUFSIZE];

	switch (target) {
	case REMOVE_GRACE_USED_SPACE_RATIO:
		snprintf(config, sizeof config,
		    "replica_check_remove_grace_used_space_ratio");
		break;
	case REMOVE_GRACE_TIME:
		snprintf(config, sizeof config,
		    "replica_check_remove_grace_time");
		break;
	case HOST_DOWN_THRESH:
		snprintf(config, sizeof config,
		    "replica_check_host_down_thresh");
		break;
	case SLEEP_TIME:
		snprintf(config, sizeof config,
		    "replica_check_sleep_time");
		break;
	case MINIMUM_INTERVAL:
		snprintf(config, sizeof config,
		    "replica_check_minimum_interval");
		break;
	default:
		e = GFARM_ERR_INVALID_ARGUMENT;
		return (e);
	}
	e = gfm_client_config_name_to_string(gfm_server,
			    config, buffer, sizeof buffer);
	if (e == GFARM_ERR_NO_ERROR)
		printf("%s\n", buffer);
	return (e);
}

static gfarm_error_t
config_set(struct gfm_connection *gfm_server, enum target target, char *value)
{
	gfarm_error_t e;
	char config[BUFSIZE];

	switch (target) {
	case REMOVE_GRACE_USED_SPACE_RATIO:
		snprintf(config, sizeof config,
		    "replica_check_remove_grace_used_space_ratio %s", value);
		break;
	case REMOVE_GRACE_TIME:
		snprintf(config, sizeof config,
		    "replica_check_remove_grace_time %s", value);
		break;
	case HOST_DOWN_THRESH:
		snprintf(config, sizeof config,
		    "replica_check_host_down_thresh %s", value);
		break;
	case SLEEP_TIME:
		snprintf(config, sizeof config,
		    "replica_check_sleep_time %s", value);
		break;
	case MINIMUM_INTERVAL:
		snprintf(config, sizeof config,
		    "replica_check_minimum_interval %s", value);
		break;
	default:
		e = GFARM_ERR_INVALID_ARGUMENT;
		return (e);
	}
	e = gfm_client_config_set_by_string(gfm_server, config);
	return (e);
}

int
main(int argc, char *argv[])
{
	gfarm_error_t e;
	int c;
	const char *path = ".";
	char *realpath = NULL;
	struct gfm_connection *gfm_server;
	enum target arg1 = STATUS1;
	enum value arg2 = STATUS2;
	char *arg2_str = NULL;
	int status;

	if (argc > 0)
		program_name = basename(argv[0]);

	while ((c = getopt(argc, argv, "P:?")) != -1) {
		switch (c) {
		case 'P':
			path = optarg;
			break;
		case '?':
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	e = gfarm_initialize(&argc, &argv);
	error_check("gfarm_initialize", e);

	if (argc < 1)
		usage();
	if (strcmp(*argv, "start") == 0 || strcmp(*argv, "enable") == 0)
		arg1 = START;
	else if (strcmp(*argv, "stop") == 0 || strcmp(*argv, "disable") == 0)
		arg1 = STOP;
	else if (strcmp(*argv, "status") == 0)
		arg1 = STATUS1;
	else if (strcmp(*argv, "remove") == 0)
		arg1 = REMOVE;
	else if (strcmp(*argv, "reduced_log") == 0)
		arg1 = REDUCED_LOG;
	else if (strcmp(*argv, "remove_grace_used_space_ratio") == 0)
		arg1 = REMOVE_GRACE_USED_SPACE_RATIO;
	else if (strcmp(*argv, "remove_grace_time") == 0)
		arg1 = REMOVE_GRACE_TIME;
	else if (strcmp(*argv, "host_down_thresh") == 0)
		arg1 = HOST_DOWN_THRESH;
	else if (strcmp(*argv, "sleep_time") == 0)
		arg1 = SLEEP_TIME;
	else if (strcmp(*argv, "minimum_interval") == 0)
		arg1 = MINIMUM_INTERVAL;
	else
		usage();

	if (arg1 == START || arg1 == STOP || arg1 == STATUS1) {
		if (argc != 1)
			usage();
	} else {
		if (argc != 2)
			usage();
		argv++;
		arg2_str = *argv;
		if (strcmp(*argv, "enable") == 0)
			arg2 = ENABLE;
		else if (strcmp(*argv, "disable") == 0)
			arg2 = DISABLE;
		else if (strcmp(*argv, "status") == 0)
			arg2 = STATUS2;
		else if (strcmp(*argv, "") == 0)
			usage();
		else
			arg2 = SET_VALUE;
	}

	if (gfarm_realpath_by_gfarm2fs(path, &realpath) == GFARM_ERR_NO_ERROR)
		path = realpath;
	e = gfm_client_connection_and_process_acquire_by_path(
		path, &gfm_server);
	error_check(path, e);

	switch (arg1) {
	case START:
		e = gfm_client_replica_check_ctrl_start(gfm_server);
		break;
	case STOP:
		e = gfm_client_replica_check_ctrl_stop(gfm_server);
		break;
	case STATUS1:
		e = gfm_client_replica_check_status_mainctrl(
		    gfm_server, &status);
		if (e == GFARM_ERR_NO_ERROR)
			printf("%s\n",
			    gfm_client_replica_check_status_string(status));
		break;
	case REMOVE:
		switch (arg2) {
		case ENABLE:
			e = gfm_client_replica_check_ctrl_remove_enable(
			    gfm_server);
			break;
		case DISABLE:
			e = gfm_client_replica_check_ctrl_remove_disable(
			    gfm_server);
			break;
		case STATUS2:
			e = gfm_client_replica_check_status_remove(
			    gfm_server, &status);
			if (e == GFARM_ERR_NO_ERROR)
				printf("%s\n",
				    gfm_client_replica_check_status_string(
					status));
			break;
		default:
			usage();
		}
		break;
	case REDUCED_LOG:
		switch (arg2) {
		case ENABLE:
			e = gfm_client_replica_check_ctrl_reduced_log_enable(
			    gfm_server);
			break;
		case DISABLE:
			e = gfm_client_replica_check_ctrl_reduced_log_disable(
			    gfm_server);
			break;
		case STATUS2:
			e = gfm_client_replica_check_status_reduced_log(
			    gfm_server, &status);
			if (e == GFARM_ERR_NO_ERROR)
				printf("%s\n",
				    gfm_client_replica_check_status_string(
					status));
			break;
		default:
			usage();
		}
		break;
	default:
		switch (arg2) {
		case STATUS2:
			e = config_get(gfm_server, arg1);
			break;
		case SET_VALUE:
			e = config_set(gfm_server, arg1, arg2_str);
			error_check(arg2_str, e);
			e = GFARM_ERR_NO_ERROR; /* error_check()ed */
			break;
		default:
			usage();
		}

	}
	error_check(path, e);

	free(realpath);
	gfm_client_connection_free(gfm_server);

	e = gfarm_terminate();
	error_check("gfarm_terminate", e);

	exit(0);
}
